program-runtime/src/compute_budget.rs

```diff
@@ -102,35 +102,54 @@ impl ComputeBudget {
-        // Compute budget instruction must be in the 1st 3 instructions (avoid
-        // nonce marker), otherwise ignored
-        for (program_id, instruction) in message.program_instructions_iter().take(3) {
+        for (i, (program_id, instruction)) in message.program_instructions_iter().enumerate() {
             if compute_budget::check_id(program_id) {
-                match try_from_slice_unchecked(&instruction.data) {
-                    Ok(ComputeBudgetInstruction::RequestUnits {
-                        units,
-                        additional_fee,
-                    }) => {
-                        self.max_units = units.min(MAX_UNITS) as u64;
-                        requested_additional_fee = additional_fee as u64;
-                    }
-                    Ok(ComputeBudgetInstruction::RequestHeapFrame(bytes)) => {
-                        if !requestable_heap_size
-                            || bytes > MAX_HEAP_FRAME_BYTES
-                            || bytes < MIN_HEAP_FRAME_BYTES as u32
-                            || bytes % 1024 != 0
-                        {
-                            return Err(error);
+                // don't include request instructions in default max calc
+                num_instructions = num_instructions.saturating_sub(1);
+
+                // Compute budget instruction must be in the 1st 3 instructions (avoid
+                // nonce marker), otherwise ignored
+                if i < 3 {
+                    match try_from_slice_unchecked(&instruction.data) {
+                        Ok(ComputeBudgetInstruction::RequestUnits {
+                            units,
+                            additional_fee,
+                        }) => {
+                            requested_units = Some(units as u64);
+                            requested_additional_fee = additional_fee as u64;
+                        }
+                        Ok(ComputeBudgetInstruction::RequestHeapFrame(bytes)) => {
+                            if !requestable_heap_size
+                                || bytes > MAX_HEAP_FRAME_BYTES
+                                || bytes < MIN_HEAP_FRAME_BYTES as u32
+                                || bytes % 1024 != 0
+                            {
+                                return Err(error);
+                            }
+                            self.heap_size = Some(bytes as usize);
                         }
-                        self.heap_size = Some(bytes as usize);
+                        _ => return Err(error),
                     }
-                    _ => return Err(error),
                 }
             }
         }
+
+        self.max_units = if default_units_per_instruction {
+            requested_units
+                .or_else(|| Some(num_instructions.saturating_mul(DEFAULT_UNITS as usize) as u64))
+        } else {
+            requested_units
+        }
+        .unwrap_or(MAX_UNITS as u64)
+        .min(MAX_UNITS as u64);
```

Comment:
should we really  clamp here? failure is always an option

Reply:
figure clamp and hope for best, max might change but in a way that would still allow the tx to succeed

Reply:
yeah, just thought it might be confusing DX